tensorboard/data/server/writer.rs

```diff
@@ -0,0 +1,126 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+//! Test helpers for writing event files.
+
+use std::io::Write;
+
+use crate::proto::tensorboard as pb;
+use crate::types::{Step, Tag, WallTime};
+
+/// Extends [`Write`] with methods for writing summary event files.
+pub trait SummaryWriteExt: Write {
+    /// Writes a TFRecord containing an `Event` proto into this writer.
+    fn write_event(&mut self, event: &pb::Event) -> std::io::Result<()> {
+        use prost::Message;
+        let mut data = Vec::new();
+        event.encode(&mut data)?;
+        crate::tf_record::TfRecord::from_data(data).write(self)?;
+        Ok(())
+    }
+
+    /// Writes a TFRecord containing a TF 1.x scalar event (`simple_value`) into this writer.
+    fn write_scalar(
+        &mut self,
+        tag: &Tag,
+        step: Step,
+        wt: WallTime,
+        value: f32,
+    ) -> std::io::Result<()> {
+        let event = pb::Event {
+            step: step.0,
+            wall_time: wt.into(),
+            what: Some(pb::event::What::Summary(pb::Summary {
+                value: vec![pb::summary::Value {
+                    tag: tag.0.clone(),
+                    value: Some(pb::summary::value::Value::SimpleValue(value)),
+                    ..Default::default()
+                }],
+                ..Default::default()
+            })),
+            ..Default::default()
+        };
+        self.write_event(&event)
+    }
+}
+
+impl<W: Write> SummaryWriteExt for W {}
```

Comment:
Here’s the secret sauce: we implement `SummaryWriteExt` for all types
`W` that implement `Write`. Since `SummaryWriteExt` only has methods
with default implementations that are implemented in terms of the
existing methods on `Write`, we don’t actually need to put anything in
this block. Thus, any caller can access these methods by just bringing
`SummaryWriteExt` into scope.